File: ios/AppContext.swift
import Foundation
import OSLog
import Supabase
import SwiftUI

enum AppRole: String, Codable, Sendable {
    case therapist
    case patient
}

enum PatientActivation: String, Codable, Sendable {
    case active
    case incomplete
}

/// Access/context for the signed-in Supabase user, as decided by the
/// `get-app-context` Edge Function. Patient-only fields are absent for
/// therapists. Entitlement is intentionally not modeled yet.
nonisolated struct AppContext: Codable, Equatable, Sendable {
    let version: Int
    let role: AppRole
    let activation: PatientActivation?
    let patientId: UUID?

    var isActivePatient: Bool {
        role == .patient && activation == .active && patientId != nil
    }

    var isIncompletePatient: Bool {
        role == .patient && activation == .incomplete
    }
}

/// Fetches `AppContext` through the shared authenticated Supabase client.
@Observable
@MainActor
final class AppContextService {
    private let client: SupabaseClient

    private(set) var current: AppContext?
    private(set) var isLoading = false
    private(set) var loadError: String?

    init(client: SupabaseClient) {
        self.client = client
    }

    func getCurrentAppContext() async throws -> AppContext {
        guard SupabaseConfig.isConfigured else { throw AuthError.notConfigured }
        isLoading = true
        loadError = nil
        defer { isLoading = false }
        do {
            let context: AppContext = try await client.functions.invoke("get-app-context")
            current = context
            AppLog.auth.info("App context received, role: \(context.role.rawValue, privacy: .public)")
            return context
        } catch {
            loadError = error.localizedDescription
            throw error
        }
    }

    func clear() {
        current = nil
        loadError = nil
        isLoading = false
    }
}

File: android/app/src/main/java/com/cbtipul/app/data/AppContext.kt
package com.cbtipul.app.data

import com.cbtipul.app.debug.InviteDebugLog
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.functions.functions
import io.ktor.client.statement.bodyAsText
import io.ktor.http.ContentType
import io.ktor.http.Headers
import io.ktor.http.HttpHeaders
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.serialization.Serializable

@Serializable
enum class AppRole {
    @kotlinx.serialization.SerialName("therapist") Therapist,
    @kotlinx.serialization.SerialName("patient") Patient,
}

@Serializable
enum class PatientActivation {
    @kotlinx.serialization.SerialName("active") Active,
    @kotlinx.serialization.SerialName("incomplete") Incomplete,
}

@Serializable
data class AppContext(
    val version: Int = 1,
    val role: AppRole,
    val activation: PatientActivation? = null,
    val patientId: String? = null,
) {
    val isActivePatient: Boolean
        get() = role == AppRole.Patient && activation == PatientActivation.Active && !patientId.isNullOrBlank()

    val isIncompletePatient: Boolean
        get() = role == AppRole.Patient && activation == PatientActivation.Incomplete
}

class AppContextRepository(private val client: SupabaseClient) {
    private val _current = MutableStateFlow<AppContext?>(null)
    val current: StateFlow<AppContext?> = _current.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    suspend fun getCurrentAppContext(): AppContext {
        if (!SupabaseConfig.isConfigured) throw IllegalStateException("not_configured")
        _isLoading.value = true
        return try {
            val http = client.functions.invoke(
                function = "get-app-context",
                headers = Headers.build {
                    append(HttpHeaders.ContentType, ContentType.Application.Json.toString())
                },
            )
            val context = EdgePayload.json.decodeFromString(AppContext.serializer(), http.bodyAsText())
            _current.value = context
            context
        } catch (error: Exception) {
            InviteDebugLog.e("get-app-context", error)
            throw error
        } finally {
            _isLoading.value = false
        }
    }

    fun clear() {
        _current.value = null
        _isLoading.value = false
    }
}
